server/admin/admin_server.go

```diff
@@ -72,12 +74,28 @@ func (s *Server) getTreeImpl(ctx context.Context, request *trillian.GetTreeReque
 func (s *Server) CreateTree(ctx context.Context, request *trillian.CreateTreeRequest) (*trillian.Tree, error) {
 	tree, err := s.createTreeImpl(ctx, request)
 	if err != nil {
-		return nil, errors.WrapError(err)
+		return nil, svrerrors.WrapError(err)
 	}
 	return tree, err
 }
 
 func (s *Server) createTreeImpl(ctx context.Context, request *trillian.CreateTreeRequest) (*trillian.Tree, error) {
+	// Test that the tree's private key exists. If it doesn't, generate it.
+	if _, err := s.registry.SignerFactory.NewSigner(ctx, request.GetTree()); err != nil {
+		if errors.ErrorCode(err) != errors.NotFound {
```

Comment:
The issue with that logic is that the `tree.PrivateKey` field currently tells the `keys.Generator` how to generate the key. For example, if `PrivateKey` was a `PEMKeyFile`, it would tell the `Generator` what password to encrypt the key with. Other key types may similarly require some metadata in order to generate a key. I suppose that isn't absolutely necessary though, as the `Generator` could probably just be hard-coded with particular settings, at the cost of some flexibility. Your proposal certainly has clearer behaviour.